src/common/walFilter/versions/recordProcessGPDB6.c

```diff
@@ -0,0 +1,503 @@
+#include "build.auto.h"
+
+#include <stdbool.h>
+#include "common/log.h"
+#include "postgres/interface/crc32.h"
+#include "recordProcessGPDB6.h"
+#include "xlogInfoGPDB6.h"
+
+#define XLR_INFO_MASK           0x0F
+#define XLOG_HEAP_OPMASK        0x70
+
+typedef uint32 CommandId;
+typedef uint16 OffsetNumber;
+typedef uint32 ForkNumber;
+
+typedef struct BkpBlock
+{
+    RelFileNode node;           /* relation containing block */
+    ForkNumber fork;            /* fork within the relation */
+    BlockNumber block;          /* block number */
+    uint16 hole_offset;         /* number of bytes before "hole" */
+    uint16 hole_length;         /* number of bytes in "hole" */
+
+    /* ACTUAL BLOCK DATA FOLLOWS AT END OF STRUCT */
+} BkpBlock;
+
+typedef struct xl_smgr_truncate
+{
+    BlockNumber blkno;
+    RelFileNode rnode;
+} xl_smgr_truncate;
+
+typedef struct BlockIdData
+{
+    uint16 bi_hi;
+    uint16 bi_lo;
+} BlockIdData;
+
+typedef struct ItemPointerData
+{
+    BlockIdData ip_blkid;
+    OffsetNumber ip_posid;
+}
+
+#ifdef __arm__
+__attribute__((packed))         /* Appropriate whack upside the head for ARM */
+#endif
+ItemPointerData;
+
+typedef struct xl_heaptid
+{
+    RelFileNode node;
+    ItemPointerData tid;        /* changed tuple id */
+} xl_heaptid;
+
+typedef struct xl_heap_new_cid
+{
+    /*
+     * store toplevel xid so we don't have to merge cids from different
+     * transactions
+     */
+    TransactionId top_xid;
+    CommandId cmin;
+    CommandId cmax;
+
+    /*
+     * don't really need the combocid since we have the actual values right in
+     * this struct, but the padding makes it free and its useful for
+     * debugging.
+     */
+    CommandId combocid;
+
+    /*
+     * Store the relfilenode/ctid pair to facilitate lookups.
+     */
+    xl_heaptid target;
+} xl_heap_new_cid;
+
+// Get RelFileNode from XLOG record.
+// Only XLOG_FPI contains RelFileNode, so the other record types are ignored.
+static bool
+getXlog(const XLogRecord *record, RelFileNode **node)
+{
+    const uint8 info = (uint8) (record->xl_info & ~XLR_INFO_MASK);
+    switch (info)
+    {
+        case XLOG_CHECKPOINT_SHUTDOWN:
+        case XLOG_CHECKPOINT_ONLINE:
+        case XLOG_NOOP:
+        case XLOG_NEXTOID:
+        case XLOG_NEXTRELFILENODE:
+        case XLOG_RESTORE_POINT:
+        case XLOG_BACKUP_END:
+        case XLOG_PARAMETER_CHANGE:
+        case XLOG_FPW_CHANGE:
+        case XLOG_END_OF_RECOVERY:
+        case XLOG_OVERWRITE_CONTRECORD:
+        case XLOG_SWITCH:
+//          ignore
+            break;
+
+        case XLOG_FPI:
+        {
+            *node = (RelFileNode *) XLogRecGetData(record);
+            return true;
+        }
+
+        default:
+            THROW_FMT(FormatError, "XLOG UNKNOWN: %d", info);
+    }
+    return false;
+}
```

Comment:
I think it would look nicer if the end of these functions looked like this:
```suggestion
    THROW_FMT(FormatError, "XLOG UNKNOWN: %d", info);
}
```

Reply:
done